File: Pre CAT/DownCasting.cpp
#include<iostream>
using namespace std;
class A{
    private:
        int x;
    public:
        A(){
            x=0;
        }
        
        A(int x){
            this->x=x;
        }
        
        virtual void print(){
            cout<<"x: "<<x<<endl;
        }
};
class B: public A{
    private:
        int a;
    public:
        B(){
            a=0;
        }

        B(int x,int a): A(x){
            this->a=a;
        }

        void print(){
            cout<<"a: "<<a<<endl;
        }
};
int main(){
    B b(5,10);
    A *a = &b;    //upcasting
    B *b1 = (B*)a;   //downcasting   B *b = &b;
    b1->print();
}

File: CDAC/Java_Team_6/Downcasting.java
class XY{
	static int x;
	int y;
	static{
		x=5;
	}
	XY(){
		y=10;
	}
	static void fun(){
		System.out.println("XY::fun()");
	}
	void test(){
		System.out.println("XY::test()");
	}
}
class YZ extends XY{
	static int a;
	int b;
	static{
		a=15;
	}
	YZ(){
		b=20;
	}
	static void fun(){
		System.out.println("YZ::fun()");
	}
	void test(){
		System.out.println("YZ::test()");
	}
	
	public static void main(String[] args){
		//YZ y = (YZ)new XY();     //Downcasting
		XY x = new YZ();
		YZ y = (YZ)x;    //  YZ y = new YZ();
		System.out.println(y.x);   //5
		System.out.println(y.y);   //10
		System.out.println(y.a);   //15
		System.out.println(y.b);   //20
		y.test();                  //YZ::test()
		y.fun();				   //YZ::fun()
	}
}